src/FileStorageBase/Network/Address.cs

```diff
@@ -0,0 +1,246 @@
+using System;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
+using Multiformats.Address;
+using Multiformats.Address.Net;
+
+namespace Neo.FileStorage.Network
+{
+    public class Address : IEquatable<Address>
+    {
+        public const string L4Protocol = "tcp";
+        private readonly Multiaddress ma;
+
+        public Address() { }
+
+        public Address(Multiaddress m)
+        {
+            ma = m;
+        }
+
+        public Address Encapsulate(Address other)
+        {
+            return new(ma.Encapsulate(other.ma));
+        }
+
+        public Address Decapsulate(Address other)
+        {
+            return new(ma.Decapsulate(other.ma));
+        }
+
+        public string ToIPAddressString()
+        {
+            return ma.ToEndPoint().ToString();
+        }
+
+        public string ToHostAddressString()
+        {
+            return DialArgs(ma);
+        }
+
+        public static Address FromString(string s)
+        {
+            Multiaddress m;
+            try
+            {
+                m = Multiaddress.Decode(s);
+            }
+            catch (Exception)
+            {
+                m = Multiaddress.Decode(MultiAddrStringFromHostAddr(s));
+            }
+            return new Address(m);
+        }
+
+        public bool Equals(Address other)
+        {
+            if (other is null) return false;
+            return other.ma.Equals(ma);
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj is null) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            if (obj is Address addr)
+                return Equals(addr);
+            return false;
+        }
+
+        public override int GetHashCode() => ma.ToString().GetHashCode();
```

Comment:
```suggestion
        public override int GetHashCode() => ma.GetHashCode();
```

Reply:
Will test this. Not sure if `MultiAddress.GetHashCode()` is implemented properly.

Reply:
If we use `ma.GetHashCode()`, Two `Network.Address` parsed from the same string will have different hash code.  Then we can't get `intersect`  between two lists.

Reply:
If `GetHashCode` collide, it should call `Equal` method